Src/WindowsServer/WindowsServer.Shared/AzureInstanceMetadataTelemetryInitializer.cs

```diff
@@ -0,0 +1,53 @@
+﻿namespace Microsoft.ApplicationInsights.WindowsServer
+{
+    using System.Threading.Tasks;
+    using Microsoft.ApplicationInsights.Channel;
+    using Microsoft.ApplicationInsights.Extensibility;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing;
+    using Microsoft.ApplicationInsights.WindowsServer.Implementation;
+
+    /// <summary>
+    /// A telemetry initializer that adds Azure specific context information to the heartbeat, if it is available.
+    /// </summary>
+    public class AzureInstanceMetadataTelemetryInitializer : ITelemetryInitializer
```

Comment:
Is this initializer meant to be automatically added to ai.config? If yes, ApplicationInsights.config.install.xdt need updated.

Reply:
@SergeyKanzhelev to confirm if this is needed.

Reply:
Considering the Initialize method is blank, does it really need to be an ITelemetryInitializer? Could it be an ITelemetryModule instead?

Reply:
@cijothomas I think I might not understand this comment… If we don’t create this initializer the heartbeat will never be updated?

@pharring I’m not aware of the difference, or the cost/benefit of choosing Module over Initializer. I need a place to do an initial ‘fire and forget’ to populate some properties into a cache. Is there performance or other considerations that makes a `TelemetryModule` more adequate? I chose `TelemetryInitializer` as that is the true nature of what I’m trying to accomplish, but I am happy to be corrected if there is a consideration I’ve missed.

Reply:
Yes, performance. The `Initialize` method of an `ITelemetryInitializer` will be called for every tracked telemetry item. Even an empty method has an overhead (it's a virtual call through the interface). On the other hand an `ITelemetryModule` has an `Initialize` method that is called just once when the module is added to the telemetry pipeline. It's more appropriate for the fire-and-forget task (move it out of the constructor into the `ITelemetryModule.Initialize` method)

Reply:
Now I look closer at what you're doing within the constructor (enumerating all previous telemetry modules), perhaps that wouldn't be appropriate because you can't ensure you're the last module to be loaded (or at least after all the other `IHeartbeatPropertyManager` instances). Indeed, even as an `ITelemetryInitializer` you're relying on an implementation detail that they are constructed after all modules are created.

Reply:
Yes, this could be problematic in the future if we change the implementation of our startup/initialization. It would be better if I could simply get 'plugged into' a module (the base SDK's `HeartbeatProvider` module, in this case) and the provider module could then initiate the heartbeat property acquisition when it was entirely ready.

Let's take this off of this PR into an issue perhaps, where I can get some further input and make necessary changes in the next beta release?

Reply:
I've discussed this with @Dmitry-Matveev and we agree it should be an `ITelemetryModule` for now. There are still some things to consider here regarding order of initialization that you brought up which we will need to handle. I am indeed leveraging knowledge of the implementation of baseSDK here, the `DiagnosticsTelemetryModule` always is added to the pipeline before all other modules, and it owns the `IHeartbeatPropertyManager` implementation.

Raised issue #834 for us to track, will be slated for beta2 timeframe.